app/donu/ExposureSession.hs

```diff
@@ -1,147 +1,156 @@
-                pokeSession esm { session = Just sess }
+
+import           Snap (MonadSnap)
+
+import           Network.WebSockets      as Sockets
+import           Network.WebSockets.Snap
+import qualified Data.ByteString.Lazy as LBS
+import qualified Data.Aeson as JS
+import           Data.Aeson ((.=), (.:))
+import qualified Data.Text.Lazy.Encoding as Text
+import qualified Control.Exception as X
+import Data.Maybe (fromMaybe)
+import qualified Language.ASKEE.Exposure.Interpreter as Exposure
+import qualified Language.ASKEE.Exposure.Syntax as Exposure
+import qualified Data.Text as Text
+import qualified Language.ASKEE.Exposure.GenLexer as Exposure
+import qualified Language.ASKEE.Exposure.GenParser as Exposure
+import Schema
+import Data.String (fromString)
+
+-- | Messages sent by the server
+data ExposureServerMessage =
+    ReadFile FilePath -- ^ Instruct the client to retrieve a file's contents
+  | WriteFile FilePath LBS.ByteString -- ^ Ask the client to write to a file
+  | Success [DonuValue] -- ^ Successfully evaluated program
+  | Failure Text.Text -- ^ Some error occurred
+
+-- | Messages sent by client
+data ExposureClientMessage =
+    RunProgram [Exposure.Stmt] -- ^ Ask the server to run some code
+  | FileContents LBS.ByteString -- ^ Provide file contents
+  | Error LBS.ByteString -- ^ This is actually an error deciphering a client message
+
+-- | The main entry point into a websocket-based Exposure session
+exposureServer :: MonadSnap m => m ()
+exposureServer = runWebSocketsSnap (acceptRequest >=> exposureServerLoop)
+
+-- | The main server loop, the toplevel of which waits for programs to try and
+-- run in the exposure interpreter.
+exposureServerLoop :: Connection -> IO ()
+exposureServerLoop c = go Exposure.initialEnv
+  where
+    go env =
+      do msg <- X.try (receiveData c)
+         case msg of
+           Right msg' -> onReceive msg' env
+           Left  ex   -> print ex >> onExcept ex
+
+    onReceive msg env =
+      case msg of
+        RunProgram prog ->
+          do (res, env') <- liftIO $ eval env prog
+             case res of
+               Left err ->
+                 do sendTextData c (Failure err)
+               Right (displays, _) ->
+                 do let display = fmap (DonuValue . Exposure.unDisplayValue) displays
+                    sendTextData c (Success display)
+             go env'
+        _ ->
+          -- This is the toplevel, so we don't expect any filecontents, for
+          -- example
+          sendTextData c (Failure "Unexpected message")
+
+    eval = Exposure.evalLoop' evr
+    evr  = Exposure.emptyEvalRead
+                { Exposure.getFileFn = readClientFile c
+                , Exposure.putFileFn = writeClientFile c
+                }
+
+    onExcept e =
+      case e of
+        CloseRequest{} -> return ()
+        ConnectionClosed -> return ()
+        _ -> return ()
```

Comment:
All three of these cases use `return ()`. Is there any point to matching on `e` here?

Reply:
No, but I'm wondering if we ought to continue the session in the remaining cases, which have to do with the server receiving garbage from the client: https://hackage.haskell.org/package/websockets-0.12.5.3/docs/Network-WebSockets.html#t:ConnectionException

I don't know websockets well enough to know when we might expect this to happen - it sounds like a serious client-side issue, so my inclination would be to close the connection. If we know that it's probably recoverable on a retry, then I'll just have the server send back a failure. If unsure, I think it's probably safe to just close the connection for now (i.e. `return ()`)

Reply:
I'd be fine with closing the connection in the event of receiving garbage from the client.